### ISSUE TITLE
`//go:wasm-module` doesn't seem to work

### ISSUE THREAD

**Issue by Mossaka** [tinygo-org/tinygo#3839]
> ... The Wasm module (which has a default value of `env`) can be specified using `//go:wasm-module <module>`.

The `//go:wasm-module <module-name>` directive supposed to change the module name to `<module-name>` given by the user. However, the produced Wasm module has a different naming scheme for the module name. 

## Code

```Go
package main

//go:wasm-module mymodule
//go:export add
func add(x, y uint32) uint32 {
	return x + y
}

// main is required for the `wasi` target, even if it isn't used.
func main() {}

```

## Wasm target

1. Build the wasm module by running `tinygo build -target=wasm -o go_example.wasm main.go`
2. Check the exported function signature by running `wasm-tools print go_example.wasm | grep export`

You will see something like

```Wasm
(export ...)
(export "main.add" (func $main.add.command_export))
(export ...)
```

## WASI target

1. Build the wasm module by running `tinygo build -target=wasm -o go_example.wasm main.go`
2. Check the exported function signature by running `wasm-tools print go_example.wasm | grep export`

You will see something like

```Wasm
(export "main.add" (func $main.add))
```

## Expected Result

```Wasm
(export "mymodule" (func $main.add))
```

## Tinygo

```Bash
tinygo version                                      
tinygo version 0.28.1 darwin/amd64 (using go version go1.20.4 and LLVM version 15.0.0)
```

**Comment by jordan-rash** [comment]
I just ran into this issue as well.  It seems no matter what module name it is given, its compiled to "package.name"

**Comment by aykevl** [comment]
Looking into it. This issue is present at least from TinyGo 0.27, probably earlier too.

**Comment by aykevl** [comment]
Here is a fix: https://github.com/tinygo-org/tinygo/issues/3845

**Comment by jordan-rash** [comment]
I pull this and it is still not working for me. 

```
└─❯ bat -r 64:66 main.go
  64   │ //go:wasm-module mymodule
  65   │ //export guestcall
  66   │ func guestcall(operationSize uint32, payloadSize uint32) bool { // nolint


└─❯ tinygo build -target=wasi -scheduler=none -o main.wasm .             

└─❯ wasm-tools parse -t main.wasm | grep 'export' | grep guestcall 
  (export "guestcall" (func $guestcall))

└─❯ tinygo version
tinygo version 0.29.0-dev darwin/arm64 (using go version go1.20.6 and LLVM version 15.0.7)
```

I would expect to see:
```
└─❯ wasm-tools parse -t main.wasm | grep 'export' | grep guestcall 
  (export "mymodule#guestcall" (func $mymodule#guestcall))
```


I wrote a quick patch (against release) last night that looked like this and it gave me what I was expecting....i suspect this isnt the best way, but with my minimial understanding of tinygo, its the best i could do 😅 

```
└─❯ git diff               
diff --git a/compiler/symbol.go b/compiler/symbol.go
index 6426604e..71ea52c3 100644
--- a/compiler/symbol.go
+++ b/compiler/symbol.go
@@ -355,6 +355,7 @@ func (c *compilerContext) parsePragmas(info *functionInfo, f *ssa.Function) {
                        } else {
                                // WebAssembly import
                                info.importName = importName
+                               info.linkName = info.module + "#" + info.importName
                        }
                }
```

**Comment by jordan-rash** [comment]
More context as to where this request is originating, using BCA tooling and given the following `.wit` file, you can see the export that `component new` is looking for -> `example:host/derp#hostrun`

Using my previous patch, i was able to achieve that by 
```
//go:wasm-module example:host/derp
//export hostrun
```

I am not advocating for my hack, just simply trying to share what I was doing when the problem came up.

```
└─❯ cat wit/host.wit                                                                                                                    
package example:host

interface derp {
  hostrun: func()
}

world host {
  import print: func(msg: string)

  export derp
}

└─❯ wasm-tools component embed --world host ./wit main.wasm -o main.embed.wasm                                                          

└─❯ wasm-tools component new main.embed.wasm --adapt wasi_snapshot_preview1=./wasi_snapshot_preview1.command.wasm -o main.component.wasm
error: failed to encode a component from module

Caused by:
    0: failed to validate exported interface `example:host/derp`
    1: module does not export required function `example:host/derp#hostrun`
```

**Comment by aykevl** [comment]
```go
//go:wasm-module mymodule
//export guestcall
func guestcall(operationSize uint32, payloadSize uint32) bool { // nolint
```
What do you expect to happen here? It doesn't make sense to put a `//go:wasm-module` on an _exported_ function, only on imported functions. You are building a module, with a particular name, and each function only has a function name (not a module name). For imported functions, yes you need to specify the module name (hence `//go:wasm-module` or now `//go:wasmimport`) but not for exported functions.

If you really want to name the function `example:host/derp#hostrun`, what about `//export example:host/derp#hostrun`? That's a strange looking function name, but if your embedder expects that then :shrug:

**Comment by jordan-rash** [comment]
> If you really want to name the function example:host/derp#hostrun, what about //export example:host/derp#hostrun?

I have to apologize, I did try this before I posted and it didn't work.  I must have been on the wrong branch because now I am getting the naming I expected.  Thank you very much. 

And yes, I agree with all your comments, but I was misunderstanding some of the build tags.  I still new to this 😅 

That said, I can confirm the fix worked for me.  Again, thank you
